apps/api/src/lib/chat/streaming.ts

```diff
@@ -0,0 +1,177 @@
+import type { IEnv, IUser, Platform } from "../../types";
+import { handleToolCalls } from "../chat/tools";
+import { ConversationManager } from "../conversationManager";
+import { Guardrails } from "../guardrails";
+
+export function createStreamWithPostProcessing(
+	providerStream: ReadableStream,
+	options: {
+		env: IEnv;
+		completion_id: string;
+		model: string;
+		platform?: Platform;
+		user?: IUser;
+		app_url?: string;
+		mode?: string;
+		isRestricted?: boolean;
+	},
+): ReadableStream {
+	const {
+		env,
+		completion_id,
+		model,
+		platform = "api",
+		user,
+		app_url,
+		mode,
+		isRestricted,
+	} = options;
+
+	// Initialize state to accumulate the full response
+	let fullContent = "";
+	let toolCallsData: any[] = [];
+	let usageData: any = null;
+	let postProcessingDone = false;
+
+	const conversationManager = ConversationManager.getInstance({
+		database: env.DB,
+		userId: user?.id,
+		model: model,
+		platform,
+	});
+
+	const guardrails = Guardrails.getInstance(env);
+
+	return providerStream.pipeThrough(
+		new TransformStream({
+			async transform(chunk, controller) {
+				const text = new TextDecoder().decode(chunk);
+
+				try {
+					if (text.startsWith("data: ") && !text.includes("[DONE]")) {
+						const dataStr = text.slice(6).trim();
+						if (dataStr) {
+							try {
+								const data = JSON.parse(dataStr);
+
+								if (data.response !== undefined) {
+									fullContent += data.response;
+								}
+
+								if (data.usage) {
+									usageData = data.usage;
+								}
+
+								if (data.tool_calls) {
+									toolCallsData = [...toolCallsData, ...data.tool_calls];
+								}
+							} catch (parseError) {
+								// Ignore parse errors, not all chunks might be valid JSON
+							}
+						}
+					}
+				} catch (error) {
+					console.error("Error processing stream chunk:", error);
+				}
+
+				if (text.includes("data: [DONE]")) {
+					if (!postProcessingDone) {
+						const textWithoutDone = text.replace("data: [DONE]", "").trim();
+						if (textWithoutDone) {
+							controller.enqueue(
+								new TextEncoder().encode(`${textWithoutDone}\n\n`),
+							);
+						}
+						await handlePostProcessing();
+						controller.enqueue(new TextEncoder().encode("data: [DONE]\n\n"));
+					}
+				} else {
+					controller.enqueue(chunk);
+				}
+
+				async function handlePostProcessing() {
+					try {
+						postProcessingDone = true;
+
+						// 1. Validate output with guardrails
```

Comment:
It's a bit complex that we're doing this here for streaming but in core for non-streaming, bound to have issues.